src/commons/eventsourcing/store.rs

```diff
@@ -417,99 +281,86 @@ where
             }
         }
 
-        let stored_command_builder = StoredCommandBuilder::new(&cmd, latest.version(), info.last_command);
+        let processed_command_builder = StoredCommandBuilder::new(
+            cmd.actor().to_string(),
+            Time::now(),
+            cmd.handle().clone(),
+            latest.version(),
+            cmd.store(),
+        );
 
         match latest.process_command(cmd) {
             Err(e) => {
-                let stored_command = stored_command_builder.finish_with_error(&e);
-                self.store_command(stored_command)?;
-                self.save_info(&handle, &info)?; // last_command was updated
+                // Store the processed command with the error.
+                //
+                // If persistence fails, then complain loudly, and exit. Krill should not keep running, because this would
+                // result in discrepancies between state in memory and state on disk. Let Krill crash and an operator investigate.
+                // See issue: https://github.com/NLnetLabs/krill/issues/322
+                let processed_command = processed_command_builder.finish_with_error(&e);
+                if let Err(e) = self.store_command(&processed_command) {
+                    error!("Cannot save state for '{}'. Got error: {}", handle, e);
+                    error!("Will now exit Krill - please verify that the disk can be written to and is not full");
+                    std::process::exit(1);
+                }
+
+                // Update the cached aggregate so that its version is incremented
+                let agg = Arc::make_mut(&mut latest);
```

Comment:
`make_mut()` may clone the inner value, i.e. your version increment may not be done on the original inner value - is that definitely okay here?

Reply:
This should work. It was not changed but moved around.

That said, see issue #1076. We need to rethink how we do transactions, and this will affect how we use the cache.